[Sol] Step02-4 - 게임로직 구현하기

Hansolkkim opened the pull request:
# 02-4) 게임로직 구현하기
## 작업 목록
- [x] 게임에 참가하는 딜러, 참가자 구현
- [x] 포커게임의 진행 시나리오를 관리해주는 PokerGame 클래스 구현
- [x] PokerGame의 메서드 동작을 확인하는 테스트 메서드 선언

---
## 질문
- 테스트에 대한 개념이 아직 제대로 잡히지 않아, 메서드 동작을 확인하는 테스트 코드를 작성할 때에도 뭔가 명확하게 그 동작을 확인하는 코드를 짰다는 생각이 들지 않았습니다. 예를 들어, 이번 Step에서 만든 PokerGame 클래스의 distributeCard() 메서드 동작을 확인하기 위해 아래와 같은 테스트 코드를 작성했습니다.
  ```swift
    func testDistributeCardInPokerGame() {
      let testPokerGame = PokerGame(numberOfGamblers: 3)
      testPokerGame.distributeCard()
        
      var countCards: Int {
          return testPokerGame.gamblers.reduce(0) { previousSum, nextGambler in
              return previousSum + nextGambler.cardDeck.count
          } + testPokerGame.dealer.cardDeck.count
      }
        
      XCTAssertTrue(countCards == testPokerGame.gameRule.numberOfCard * (testPokerGame.gamblers.count + 1), "제대로 카드가 나눠지지 않았습니다.")
    }
  ```
  위의 테스트 코드는 `카드가 나눠졌다면, 참가자와 딜러의 카드의 개수의 총합이 내가 생각한 개수와 같아야한다.` 라는 생각으로 작성한 테스트 코드 입니다.
  카드가 셔플된 후에 나눠지기 때문에, 누가 어떤 카드를 갖게되나에 대해서 테스트하는건 불가능했고, 고민 끝에 이런 코드를 작성하게 되었습니다. 이런 방식으로 테스트코드를 작성하는 것도 올바른 방법일까요?

- Mutable, Imutable한 상황을 구별하는게 힘들었습니다. 
  만약 PokerGame을 클래스가 아닌 구조체로 구현한다면, 딜러(구조체) 인스턴스 타입인 프로퍼티 dealer의 mutating한 메서드 `pickCard()`을 호출하는 것이 안되었고, 
  PokerGame을 클래스로 구현하고, 딜러에 해당하는 프로퍼티 dealer를 let으로 선언해주는 경우에도 `pickCard()` 메서드 호출이 불가능했습니다. 
  이를 해결하기 위해 Dealer, Gambler, PokerGame을 모두 클래스로 구현했습니다.
  혹시 이런 내용에 대해서 공부할 수 있는 키워드를 알려주실 수 있으신가요?

Hansolkkim committed:
Fix: 코드 리뷰에 따른 코드 수정

Hansolkkim committed:
Fix: shuffle() 메서드 오류를 줄이기 위한 코드 추가

코드 리뷰에 JK께 단언했던 부분에 오류가 있음을 발견해서 오류를 줄이기 위한 코드를 추가함

Hansolkkim committed:
Docs: README.md에 Step3 진행사항, 추가학습거리, 코드리뷰 피드백 내용 추가

Hansolkkim committed:
Fix: 불필요한 부분 삭제 및 수정, 주석 추가

Hansolkkim committed:
Feat: 딜러와 게임참가자가 채택해야할 Gamer 프로토콜 선언

Hansolkkim committed:
Feat: 딜러를 구현할 Dealer 구조체 선언

Hansolkkim committed:
Feat: 게임 참가자를 구현한 Gambler 구조체 선언

Hansolkkim committed:
Feat: 포커 게임을 진행하면서 필요한 add(_:) 메서드를 CardDeck 구조체에 선언

Hansolkkim committed:
Feat: Dealer 구조체, Gambler 구조체, CardDeck 구조체가 CustomStringConvertible 프로토콜을 채택

Hansolkkim committed:
Feat: 포커게임을 구현할 PokerGame 클래스 선언

Hansolkkim committed:
Feat: 포커게임 진행사항을 프린트해주는 gamePrgram() 함수 선언

Hansolkkim committed:
Rename: 파일 구조 변경

Model에 해당하는 Swift 파일들을 Model 폴더로 이동

Hansolkkim committed:
Rename: PokerGame 클래스의 handOut() 메서드 이름을 distibuteCard() 로 변경

Hansolkkim committed:
Fix: 이름이 2~5글자 이내가 아닌 이름 제거

Hansolkkim committed:
Feat: 게임 시작 후 출력을 위한 함수를 ViewController 클래스의 viewDidLoad() 메서드에서 호출

Hansolkkim committed:
Test: PokerGame 클래스의 distributeCard() 메서드 테스트 메서드 추가

Hansolkkim committed:
Remove: 불필요한 Gamer 프로토콜 제거

Hansolkkim committed:
Remove: 불필요한 파일 및 코드 제거

Hansolkkim committed:
Fix: 구조체로 선언했던 Dealer, Gambler를 클래스로 변경

godrm commented:
테스트는 어떤 형태로든 입력 상황에 대해 출력값을 확인할 수 있으면 테스트라고 할 수있죠
다만 지금처럼 여러 메소드나 여러 객체가 서로 협력해서 사용해야 하는 경우는 어느 객체가 잘못되거나 오동작하는지 복잡적이라서 찾기 어려울 수 있습니다. 그래서 `단위 테스트`라고 한 모듈/타입의 한 기능만 테스트하는 것을 우선적으로 시도하게 되죠
지금처럼 여러 메소드를 호출하는 경우는 통합 테스트라고 부르기도 합니다. 
그리고 테스트할 때 셔플 자체는 테스트하기 어렵기 때문에, 순서를 보장할 수 있는 상태로 만들어서 비교하는 방식을 많이 씁니다. 

객체 관련해서는 곧 다시 수업에서도 언급할 예정입니다.

godrm commented on PockerGameApp/PockerGameApp/Model/Card.swift:
```diff
@@ -57,11 +57,7 @@ extension Card: CustomStringConvertible {
-        } else {
-            return false
-        }
+        return lhs.number == rhs.number && lhs.symbol == rhs.symbol ? true : false
```
이전에 언급했던 것 같은 데 `return lhs.number == rhs.number && lhs.symbol == rhs.symbol` 이 자체가 true/false를 리턴합니다. 굳이 값을 할당하지 않아도 됩니다

Hansolkkim replied:
> 이전에 언급했던 것 같은 데 `return lhs.number == rhs.number && lhs.symbol == rhs.symbol` 이 자체가 true/false를 리턴합니다. 굳이 값을 할당하지 않아도 됩니다

아, 제가 그때 리뷰해주셨었던 내용을 잘못 이해했던것 같습니다.. 수정하도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/CardDeck.swift:
```diff
@@ -18,19 +18,15 @@ struct CardDeck {
-        
-        self.deck = deck
-        self.originDeck = deck
+        self.originDeck = self.deck
```
이런 경우는 프로퍼티로 빈 배열이 있고, 추가하게 되니까 괜찮은데
아래 또 다른 생성자 init(with:) 경우는 이미 빈 배열이 추가된 경우에 self.deck에 새로운 배열이 들어가니까 미리 만들지 않아도 될 수 있겠네요
별건 아니지만 프로퍼티를 literal하게 초기값을 주는 것과 init()에서 초기값을 주는 것도 기준을 잡아보세요.

Hansolkkim replied:
> 이런 경우는 프로퍼티로 빈 배열이 있고, 추가하게 되니까 괜찮은데
> 아래 또 다른 생성자 init(with:) 경우는 이미 빈 배열이 추가된 경우에 self.deck에 새로운 배열이 들어가니까 미리 만들지 않아도 될 수 있겠네요
> 별건 아니지만 프로퍼티를 literal하게 초기값을 주는 것과 init()에서 초기값을 주는 것도 기준을 잡아보세요.

저번 Step에서 init() 메서드 내에 새로운 `deck` 이라는 변수를 만들어 해당 변수에 카드들을 넣은 후에  deck을 `self.deck` 에 저장하는 방법을 사용했을 때 질문을 받았었는데, 해당 방법이 불필요하지 않냐는 말씀으로 이해하고 일부분만 수정하다보니 기준이 흐트러졌던것 같습니다.. 수정하도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/CardDeck.swift:
```diff
@@ -40,23 +36,33 @@ struct CardDeck {
+            self.deck.swapAt(indexToSwap1, indexToSwap2)
+        }
+        if self.deck[count-1] == self.originDeck[count-1] {
+            self.deck.swapAt(count-1, count-2)
```
이 구문을 보니 count가 1인 경우는 어떻게 되나요?

Hansolkkim replied:
> 이 구문을 보니 count가 1인 경우는 어떻게 되나요?

인덱스 초과로 에러가 발생하네요,, 해당 메서드실행이 되면 무조건 순서가 하나 이상은 바뀌도록 설정하기 위해 추가한 부분인데 고민이 부족했던것 같습니다.. 에러 안나도록 수정해보겠습니다!

godrm commented on PockerGameApp/PockerGameApp/Model/CardDeck.swift:
```diff
@@ -40,23 +36,33 @@ struct CardDeck {
-        else {
-            return self.deck.removeLast()
-        }
+        return self.deck.isEmpty ? nil : self.deck.removeLast()
```
if 구문과 삼항 연산자는 어떤 차이가 있나요? 
기존에 if를 왜 삼항연산자로 바꾸었는지 궁금합니다. 
한 줄로 쓰면 오히려 변화한 부분만 찾기 어려운 경우도 있습니다

Hansolkkim replied:
> if 구문과 삼항 연산자는 어떤 차이가 있나요?
> 기존에 if를 왜 삼항연산자로 바꾸었는지 궁금합니다.
> 한 줄로 쓰면 오히려 변화한 부분만 찾기 어려운 경우도 있습니다

if 구문과 삼항 연산자는 정할 수 있는조건의 개수가 차이가 난다고 생각하여, 한가지 조건에 대한 출력값 변화를 좀 더 간결하게 표시하려고 삼항연산자를 사용했습니다. 해당 부분을 Optional값도 리턴 가능한 `popLast()` 메서드를 사용해 변경하겠습니다! 그런데 `변화한 부분만 찾기 어려운 경우` 가 어떤 경우를 의미하는지 이해가 잘 가지 않는데 말씀 좀 부탁드려도 되나요?

godrm commented on PockerGameApp/PockerGameApp/Model/Dealer.swift:
```diff
@@ -0,0 +1,32 @@
+class Dealer {
+    
+    var wholeDeck = CardDeck()
+    var cardDeck = CardDeck.init(with: [])
```
딜러가 나눠줄 카드덱과 받는 카드덱이 꼭 같은 타입이어야 하나요?

Hansolkkim replied:
> 딜러가 나눠줄 카드덱과 받는 카드덱이 꼭 같은 타입이어야 하나요?

`cardDeck` 프로퍼티가 굳이 CardDeck 타입이어야 할 필요가 없다고 생각들어서 Card Array로 수정했습니다. 그런데 오늘 강의를 듣고 Card Array도 새로운 타입으로 생성하여 수정해보도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/Dealer.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    public func pickCard() -> Card? {
+        wholeDeck.shuffle()
+        return wholeDeck.removeOne()
```
이렇다는 얘기는 카드를 나눠줄 때마다 덱을 다시 섞는걸까요?

Hansolkkim replied:
> 이렇다는 얘기는 카드를 나눠줄 때마다 덱을 다시 섞는걸까요?

생각해보니 실제 카드 배분하는 방식과 다르네요.. 카드를 나눠주기 전에만 섞도록 수정하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/Dealer.swift:
```diff
@@ -0,0 +1,32 @@
+
+extension Dealer: CustomStringConvertible {
+    var description: String {
+        return "딜러 \(self.cardDeck)"
```
내부 속성을 description으로 제공하는 것은 다분히 디버깅을 위한 목적이 큽니다.
혹시라도 description을 로직에서 비교한다거나, 출력하기 위한 용도로 쓰는 것은 지양하세요

Hansolkkim replied:
> 내부 속성을 description으로 제공하는 것은 다분히 디버깅을 위한 목적이 큽니다.
> 혹시라도 description을 로직에서 비교한다거나, 출력하기 위한 용도로 쓰는 것은 지양하세요

테스트 코드 작성 전에 카드가 잘 배분됐는지 확인하려고 추가했던 extension들이었는데 못 지웠네요.. 필요없는 CustomStringConvertible 채택 코드들은 지우도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/Gambler.swift:
```diff
@@ -0,0 +1,29 @@
+class Gambler {
+    
+    let name: String
+    var cardDeck = CardDeck.init(with: [])
```
Gambler의 cardDeck은 private 이 아니라서 외부에서 몰래 바꿔치지 할 수도 있는거네요?
항상 속성은 private을 우선적으로 고려하시기 바랍니다

Hansolkkim replied:
> Gambler의 cardDeck은 private 이 아니라서 외부에서 몰래 바꿔치지 할 수도 있는거네요?
> 항상 속성은 private을 우선적으로 고려하시기 바랍니다

Gambler와 Dealer 클래스 프로퍼티들을 private으로 수정하도록 하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/Gambler.swift:
```diff
@@ -0,0 +1,29 @@
+    var cardDeck = CardDeck.init(with: [])
+    
+    init(name: String) {
+        self.name = name
```
뭔가 이름에 대한 요구사항이 빠진거 같네요

Hansolkkim replied:
> 뭔가 이름에 대한 요구사항이 빠진거 같네요

Gambler 인스턴스를 생성할 때, 제가 생성했던 이름 배열 내에서만 이름을 사용하려고 생각해서 미처 요구사항을 따지지 못 했습니다.. 적절한 방법을 생각해서 수정하도록 하겠습니다!

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+        var nameArray = ["Sol", "Jee", "Eddy", "Jed", "Dale"]
+        var gamblers = [Gambler]()
+        for _ in 0..<numberOfGamblers{
+            let newGambler = Gambler(name: nameArray.remove(at: (0..<nameArray.count).randomElement() ?? 0))
```
속성을 구할 때도 구문 내부에 이렇게 표현식을 넣는 것은 읽기 어렵습니다. 
분리해도 좋은 표현 같습니다.

Hansolkkim replied:
> 속성을 구할 때도 구문 내부에 이렇게 표현식을 넣는 것은 읽기 어렵습니다.
> 분리해도 좋은 표현 같습니다.

```swift
let pickedIndex = (0..<nameArray.count).randomElement() ?? 0
let pickedName = nameArray.remove(at: pickedIndex)
let newGambler = Gambler(name: pickedName)
```
위의 코드처럼 분리해서 표현식을 나누었습니다! 감사합니다

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+        }
+        
+        self.dealer = Dealer()
+        self.gamblers = gamblers
```
gamblers라는 지역 변수가 항상 필요한가요? 프로퍼티에 처리해도 되는게 아닐까요?

Hansolkkim replied:
> gamblers라는 지역 변수가 항상 필요한가요? 프로퍼티에 처리해도 되는게 아닐까요?

여기서 지역변수를 두면서 gamblers를 초기화한 이유는, gamblers라는 프로퍼티가 상수로 선언되는게 PokerGame 클래스 내에서 더 적합하지 않을까 하는 생각에 이런 방식을 사용했습니다.
gambler라는 프로퍼티를 처음 선언할 때 초기화해주지 않은 상태라면 init() 메서드에서 위 같은 방법 외에는 생각이 나지 않아 이런 방법을 사용했습니다. 따로 방법이 있는지 궁금합니다..

godrm replied:
아 let이라서 한 번만 지정하려고 그랬다는 의미인가요?
그렇죠 명단을 한 번만 지정하려면 다 설정을 해서 넘겨줘야겠네요
다른 방법이 있을 수도 있는데, 별차이 없을꺼 같습니다

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+        
+        self.dealer = Dealer()
+        self.gamblers = gamblers
+        self.gameRule = .sevenCardStud
```
numberOfGamblers 값은 init할 때 받는데 rule은 고정인가 보내요? 별도로 고정한 이유가 있나요?

Hansolkkim replied:
> numberOfGamblers 값은 init할 때 받는데 rule은 고정인가 보내요? 별도로 고정한 이유가 있나요?

카드 개수에 대한 내용을 나중에 추가해서 미흡하게 수정됐던것 같습니다. rule도 init() 메서드 매개변수로 추가하도록 하겠습니다!

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+        distributeCard()
+    }
+    
+    public func distributeCard() {
```
gameScenario() 와 distributeCard() 가 같은 거라면 왜 구분했는지 모르겠습니다. 
내부 함수라면 private func로 선언해주면 되지 않을까요?

Hansolkkim replied:
> gameScenario() 와 distributeCard() 가 같은 거라면 왜 구분했는지 모르겠습니다.
> 내부 함수라면 private func로 선언해주면 되지 않을까요?

수정 전에 `gameScenario()` 메서드에서 카드의 수가 부족하면 `distributeCard()`가 호출되지 않도록 설정했다가, 해당 메서드가 테스트하기에 맞지 않는 것 같아서 수정을 했는데 distributeCard() 메서드와 같은 거로 해놓고 지우지 않았네요,, 지우도록 하겠습니다!

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+            for index in 0..<gamblers.count {
+                gamblers[index].getCard(self.dealer.pickCard()!)
+            }
+            dealer.getCard(self.dealer.pickCard()!)
```
옵셔널은 강제 언래핑을 지양하시고, 안전하게 처리하는 습관을 가지세요

Hansolkkim replied:
> 옵셔널은 강제 언래핑을 지양하시고, 안전하게 처리하는 습관을 가지세요

네, `guard let` 구문으로 언래핑하도록 수정하겠습니다.

godrm commented on PockerGameApp/PockerGameApp/Model/PockerGame.swift:
```diff
@@ -0,0 +1,54 @@
+    public func distributeCard() {
+        for _ in 0..<gameRule.numberOfCard {
+            for index in 0..<gamblers.count {
+                gamblers[index].getCard(self.dealer.pickCard()!)
```
카드가 부족하면 바로 죽을꺼 같은 로직인데요? 계속 실행이 가능하던가요?

Hansolkkim replied:
> 카드가 부족하면 바로 죽을꺼 같은 로직인데요? 계속 실행이 가능하던가요?

프로그래밍 요구사항에 맞는 동작을 하도록(카드가 부족하면 프로그램이 끝나도록) 위에 작성했던 `gameScenario()` 메서드를 이용했었는데, 해당 메서드 내에 카드 부족할땐 distributeCard()가 동작되지 않도록 했었습니다.
그런데 테스트하기에 맞지 않는 것 같아 해당 내용을 지웠는데 distributeCard() 메서드 내에 수정하는걸 깜빡했네요,,
함수 코드 시작 부분에 guard 구문을 추가하여 죽지 않도록 설정하겠습니다.
```swift
private func distributeCard() {
    guard dealer.countRemainingCards >= gameRule.numberOfCard * (gamblers.count + 1) else { return }
    dealer.shuffleWholeDeck()
    for _ in 0..<gameRule.numberOfCard {
       ....
    }
}
```

godrm reviewed (changes requested):
다음 피드백을 확인해보시고, 딜러, 참가자, 게임 역할을 개선해보세요

Hansolkkim committed:
Fix: 코드리뷰에 따른 수정

불필요한 값 할당 부분 삭제
동일한 프로퍼티는 init() 메서드가 여러 개더라도 초기화하는 방법을 하나의 기준으로 처리
shuffle() 메서드에서 발생할 수 있는 오류 처리하도록 수정
removeOne() 메서드에서 불필요한 삼항 연산자 제거
불필요한 extension 제거
Dealer 클래스의 cardDeck 프로퍼티 타입 수정
Dealer 클래스의 wholeDeck 프로퍼티의 접근 제어자 추가 및 그에 따른 shuffleWholeDeck() 메서드 추가
Gambler 클래스 init() 메서드 내 name 프로퍼티에 대한 조건 코드 추가
PokerGame 클래스 init() 메서드 gameRule 프로퍼티에 대한 매개변수 추가
속성 구할 때 가독성이 떨어지는 표현식 분리
PokerGame 클래스의 distribute() 메서드에서 오류 처리하도록 메서드 내용 수정

Hansolkkim commented:
수정완료했습니다.

godrm commented:
개별 항목에도 어떤 식으로 해결했는지 댓글을 달아주는 게 좋습니다.
이전과 달리 달라진 부분을 제가 다 기억하고, 맥락상 이런저런 것을 다 확인하지 못하고 넘어갈 수도 있으니까요

godrm commented on PockerGameApp/PockerGameApp/Model/Gambler.swift:
```diff
@@ -10,20 +10,31 @@ import Foundation
-    var description: String {
-        return "\(self.name) \(self.cardDeck)"
+    public func getCard(_ card: Card) {
+        self.cardDeck.append(card)
```
getCard()는 Gambler카드를 가져오는 걸까요? 넣어주는걸까요?
보통 get-은 객체 내부에서 값을 가져올 때 사용하는 편이라서 동작이 헷가리는 이름 같습니다.

Hansolkkim replied:
> getCard()는 Gambler카드를 가져오는 걸까요? 넣어주는걸까요?
> 보통 get-은 객체 내부에서 값을 가져올 때 사용하는 편이라서 동작이 헷가리는 이름 같습니다.

딜러로부터 카드를 받는 메서드를 의미하면서 Gambler가 좀 더 주체적인 상황으로 가정하고 이름을 지었는데, 고민이 조금 미흡했던것 같습니다.
`receiveCard(_:)` 로 수정하도록 하겠습니다.
